tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+builder method. If you choose a descriptive method name, this can already
+explain what the stage or segment does, without the reader having to see the
+internal workings.
+
+```php
+public static function groupAndComputeStatistics(
+    stdClass $_id,
+    Expression\ResolvesToDouble $price,
+): GroupStage {
+    return Stage::group(
+        _id: $_id,
+        lowest: Accumulator::min($price),
+        highest: Accumulator::max($price),
+        average: Accumulator::avg($price),
+        count: Accumulator::sum(1),
+    );
+}
+```
+
+When extracting logic, consider if the method can be reused elsewhere. In this
+case, by keeping the `_id` field for the `$group` stage along with the price
+field as a parameter, we can reuse this builder method in a different pipeline. 
+
+```php
+$reportDate = Expression::dateFieldPath('reportDate');
+$price = Expression::doubleFieldPath('price');
+
+self::groupAndComputeStatistics(
+    _id: object(
+        year: Expression::year($reportDate),
+        month: Expression::month($reportDate),
+        fuelType: Expression::fieldPath('fuelType'),
+        brand: Expression::fieldPath('station.brand'),
+    ),
+    price: $price,
+);
+```
+
+### Extract multiple stages
+
+We've now refactored the first pipeline stage, but the `$group` stage below is
+also relatively complex to read. To make matters worse, this pipeline stage
+works together with the `$addFields` stage below: `$group` assembles a list of
+fuel types with their prices, which is then converted to an object in
+`$addFields`. Ideally, we want to hide this implementation detail and extract
+both stages together.
+
+To do so, we once again extract a factory method, except that this time we'll be
+returning a `Pipeline` instance:
+
+```php
+public static function groupAndAssembleFuelTypePriceObject(
+    stdClass $_id,
+    Expression\ResolvesToString $fuelType,
+    Expression\ResolvesToInt $count,
+    Expression\ResolvesToDouble $lowest,
+    Expression\ResolvesToDouble $highest,
+    Expression\ResolvesToDouble $average,
+): Pipeline {
+    return new Pipeline(
+        Stage::group(
+            _id: $_id,
+            count: Accumulator::sum($count),
+            prices: Accumulator::push(
+                object(
+                    k: $fuelType,
+                    v: object(
+                        lowest: $lowest,
+                        highest: $highest,
+                        average: $average,
+                        range: Expression::subtract(
+                            $highest,
+                            $lowest,
+                        ),
+                    )
+                )
+            )
+        ),
+        Stage::addFields(
+            prices: Expression::arrayToObject(
+                Expression::arrayFieldPath('prices'),
+            ),
+        ),
+    );
+}
+```
+
+By once again keeping fields as parameters, we keep the method flexible and
```

Comment:
```suggestion
By once again keeping fields as parameters, the method remains flexible and we
```

Optional rewording to avoid repetition of "keep".